independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Injection.java

```diff
@@ -296,16 +310,78 @@ private static List<AnnotationInstance> getAllInjectionPoints(BeanDeployment bea
                 }
             }
         }
+
         for (MethodInfo method : beanClass.methods()) {
             if (skipConstructors && method.name().equals(Methods.INIT)) {
                 continue;
             }
-            AnnotationInstance inject = beanDeployment.getAnnotation(method, name);
+
+            MethodOverrideKey key = new MethodOverrideKey(method);
+            if (seenMethods.contains(key)) {
+                continue;
+            }
+            seenMethods.add(key);
+
+            AnnotationInstance inject = beanDeployment.getAnnotation(method, annotationName);
             if (inject != null) {
-                injectAnnotations.add(inject);
+                if (Modifier.isStatic(method.flags())) {
+                    LOGGER.warn("An initializer method must be non-static - ignoring: "
+                            + method.declaringClass().name() + "#"
+                            + method.name() + "()");
+                } else {
+                    injectAnnotations.add(inject);
+                }
             }
         }
+
         return injectAnnotations;
     }
 
+    // ---
+
+    // this is not entirely precise (according to the JLS), but it's better than Methods.MethodKey
+    // we should eventually unify this somewhere, ideally in Jandex
+    static class MethodOverrideKey {
+        private final String name;
+        private final List<DotName> params;
+        private final DotName returnType;
+
+        MethodOverrideKey(MethodInfo method) {
+            if (Modifier.isPrivate(method.flags())) {
+                // private methods cannot be overridden
+                this.name = method.declaringClass().name() + "|" + method.name();
```

Comment:
It wouldn't hurt to explain what the `name` represents in this case...

Reply:
I'll add a comment to the code, but this also answers the question above -- this is better than `MethodKey`, because it treats visibility correctly (or maybe almost correctly :-); to be fair, I haven't done extensive testing). Only `public` and `protected` methods may be overridden anywhere in the inheritance hierarchy; package-private methods may only be overridden in the same package, and `private` methods may not be overridden at all.

So if a subclass and its superclass both contain a `private` method with the same name and signature, they are 2 different methods and one does _not_ override the other. Same with package-private methods, if they appear in different packages.

Reply:
Done.